### ISSUE TITLE
[Bug] Seeded sampling silently selects token 0 for an empty distribution

### ISSUE THREAD

**Issue by morluto** [sgl-project/sglang#33697]
## Describe the bug

Deterministic seeded sampling silently selects token ID 0 when a categorical distribution has no valid token.

`multinomial_with_seed()` adds deterministic Gumbel noise to log probabilities and returns `argmax`. If every log probability in a row is `-inf`, every resulting score remains `-inf`, and `argmax` returns index 0.

This differs from unseeded `torch.multinomial`, which rejects an all-zero distribution.

## Reproduction

```python
import torch

from sglang.srt.layers.sampler import sampling_from_probs_torch

probs = torch.zeros((1, 4), dtype=torch.float32, device="cuda")
seeds = torch.tensor([42], dtype=torch.int64, device="cuda")
positions = torch.tensor([7], dtype=torch.int64, device="cuda")

sampling_from_probs_torch(probs, seeds, positions)
```

The returned token ID is 0 even though the distribution has zero probability mass.

## Expected behavior

An empty distribution must not produce a token ID. It should result in a clear, request-scoped failure without slowing valid sampling.

## Design constraint

The straightforward host-visible check is unsuitable for the sampling hot path:

```python
torch._check_tensor_all(torch.isfinite(logprobs).any(dim=-1), ...)
```

In current PyTorch, this reaches `cond._is_all_true().item()`, adding an unconditional GPU-to-CPU synchronization. Placing the check inside the compiled function also causes a graph break.

The asynchronous alternative avoids host synchronization but changes the failure contract:

```mermaid
flowchart TD
    A[All logits are -inf] --> B{Validation strategy}
    B --> C[Host-visible tensor check]
    C --> D[Clear RuntimeError]
    D --> E[Unconditional GPU-to-CPU synchronization]
    B --> F[Device-side async assertion]
    F --> G[No hot-path host sync]
    G --> H[Delayed device assertion and unusable CUDA context]
```

A device-side assertion is useful for diagnosing a fatal invariant, but it is not a recoverable request error.

## Decision needed

Where should invalid-distribution status become host-visible?

Possible directions are:

1. Propagate an invalid-row flag alongside sampled token IDs and inspect it at an existing scheduler synchronization boundary.
2. Validate at grammar or custom-logit-processor ownership boundaries, provided every path that can empty a distribution is covered.
3. Treat the condition as a fatal invariant and use an asynchronous assertion, explicitly accepting process-level failure semantics.

Success criteria:

- An empty seeded distribution never returns token ID 0.
- Valid seeded sampling adds no unconditional GPU-to-CPU synchronization.
- The compiled hash/Gumbel kernel remains intact.
- The failure is either request-scoped and clear, or explicitly defined as fatal.

**Comment by morluto** [comment]
Before implementing this, I’d like to confirm ownership of the device-side per-request validity flag.

The current code suggests two plausible carriers:

1. `LogitsProcessorOutput`: the sampler already attaches optional sampling-mask/logprob metadata here.
2. `GenerationBatchResult`: this flag is an unconditional post-sampling outcome consumed by the scheduler, and this object already owns overlap-scheduling state plus asynchronous D2H in `copy_to_cpu()`.

I currently lean toward adding an optional device tensor such as `invalid_sampling_mask` to `GenerationBatchResult`, populated from the sampler result and copied beside `next_token_ids`. The scheduler could inspect it after the existing `copy_done` synchronization, abort only invalid requests, and never commit the internal placeholder token. That keeps valid sampling free of a new host synchronization.

This would require either returning `(next_token_ids, invalid_mask)` from the sampler/model-runner path or introducing a small typed sampler result. The alternative is to attach the flag to `LogitsProcessorOutput`, which is less plumbing but mixes an unconditional scheduler control signal with optional logits/logprob metadata.

Would `GenerationBatchResult` be the preferred carrier here, or would maintainers rather keep all sampler-produced metadata on `LogitsProcessorOutput`? Once confirmed, I can send a focused PR covering mixed valid/invalid rows, overlap scheduling, and PP transport.

**Comment by morluto** [comment]
Correction after tracing the complete production call path: the direct reproduction does not establish a reachable request-level bug.

`sampling_from_probs_torch()` has only one production caller, and that caller supplies `softmax(logits)`. With at least one finite logit, this row has positive mass. Validated parameters also preserve at least one token: `top_k >= 1`, `top_p > 0` always retains the first sorted token, and `min_p <= 1` retains the maximum-probability token. The all-zero tensor in the reproduction bypasses those invariants by invoking the internal helper directly.

An empty/non-finite row would therefore require a separate upstream invariant failure, such as a custom logit processor or invalid grammar state masking every token, or a model producing only non-finite logits. Request-scoped handling for those producers may still be worthwhile, but it is a different and broader contract.

Given this trace, propagating an invalid-row flag through `GenerationBatchResult`, overlap relay, PP transport, and both result processors would be disproportionate for the demonstrated behavior. I am not preparing that PR without evidence of a reachable producer or maintainer direction that this defensive invariant should become recoverable scheduler state.

Recommended reclassification: direct internal-helper edge case / defensive invariant, not a confirmed production sampling bug.

**Comment by morluto** [comment]
Closing based on the corrected call-path analysis above: validated production sampling cannot construct the reproduced all-zero probability row, and adding cross-layer recovery plumbing would be disproportionate without a reachable producer. A separate issue can cover request-scoped handling if an upstream grammar, custom-processor, or model-logit producer is demonstrated.
